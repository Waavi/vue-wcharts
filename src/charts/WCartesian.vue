<script>
import VueTypes from 'vue-types'
import scaleLinear from 'd3-scale/src/linear'
import stack from 'd3-shape/src/stack'
import stackOffsetDiverging from 'd3-shape/src/offset/diverging'
import { noop } from 'lodash'
import { bound } from '../utils/maths'

import chartMixin from '../mixins/chart'

export default {
    name: 'WCartesian',
    mixins: [chartMixin],
    props: {
        stacked: VueTypes.bool.def(false),
        scatter: VueTypes.bool.def(false),
        bound: VueTypes.array.def([]),
        xBound: VueTypes.array.def([]),
        gap: VueTypes.oneOfType([VueTypes.number, VueTypes.arrayOf(VueTypes.number).def([0, 0, 0, 0])]).def(0),
    },
    data () {
        return {
            axis: {
                x: {
                    datakey: null,
                    name: null,
                },
                y: {
                    datakey: null,
                    name: null,
                },
                z: {
                    datakey: null,
                    name: null,
                    range: null,
                },
            },
        }
    },
    computed: {
        // yScale with scaleLinear of d3
        // ref: https://github.com/d3/d3-scale#_continuous
        yScale () {
            return scaleLinear()
                .domain([this.bounds.min, this.bounds.max])
                .range([this.canvas.y1 - this.padding[2], this.canvas.y0 + this.padding[0]])
        },
        // xScale with scaleLinear of d3
        // ref: https://github.com/d3/d3-scale#_continuous
        xScale () {
            const domain = this.scatter ? [this.xBounds.min, this.xBounds.max] : [0, this.dataset.length - 1]
            return scaleLinear()
                .domain(domain)
                .range([this.canvas.x0 + this.padding[3], this.canvas.x1 - this.padding[1]])
        },
        // zScale calculate like Recharts
        zScale () {
            if (this.scatter && this.axis.z.datakey) {
                const [rangeMin, rangeMax] = this.axis.z.range
                const { min: boundMin, max: boundMax } = this.zBounds
                return val => Math.sqrt((((val - boundMin) / (boundMax - boundMin)) * (rangeMax - rangeMin) + rangeMin) / Math.PI)
            }
            return val => val
        },
        // bounds
        bounds () {
            if (this.datakeys.length || this.axis.y.datakey) {
                const [boundMin, boundMax] = this.bound
                return {
                    min: this.getBound(boundMin),
                    max: this.getBound(boundMax, 'max'),
                }
            }
            return {
                max: 0,
                min: 0,
            }
        },
        xBounds () {
            if (this.scatter && this.axis.x.datakey) {
                const [boundMin, boundMax] = this.xBound
                return {
                    min: this.getBound(boundMin, 'min', 'x'),
                    max: this.getBound(boundMax, 'max', 'x'),
                }
            }
            return {
                max: 0,
                min: 0,
            }
        },
        zBounds () {
            if (this.scatter && this.axis.z.datakey) {
                const values = this.dataset.map(d => d[this.axis.z.datakey])
                return {
                    min: Math.min(...values),
                    max: Math.max(...values),
                }
            }
            return {
                max: 0,
                min: 0,
            }
        },
        // Generate padding by array or number.
        // return [10, 10, 10, 10]
        padding () {
            const gap = Array.isArray(this.gap) ? this.gap : Array(4).fill(this.gap)
            return gap.map((item, index) => item + this.offset[index])
        },
        // Offset of bars
        // return [10, 10, 10, 10]
        offset () {
            const { barAllWidth, barIds = [] } = this.snap
            const gap = Array(4).fill(0)

            if (barIds.length) {
                // Checked if width of bars it is higher than canvas width
                const margin = this.width <= barAllWidth * (this.dataset || []).length ? this.width % barAllWidth : barAllWidth

                gap[1] = margin
                gap[3] = margin
                return gap
            }

            return gap
        },
        // Return stacked data or data dataset without transform, of datakeys and dataset props
        // ref: https://github.com/d3/d3-shape#stacks
        curData () {
            const datakeys = this.stacked ? Object.values((this.snap || {}).barsByDatakeys || []) : this.datakeys
            const stacked = this.stacked ? stackOffsetDiverging : noop

            return stack()
                .keys(datakeys)
                .offset(stacked)(this.dataset)
        },
    },
    methods: {
        // Calc min/max bound values
        getBound (val, type = 'min', axis = 'y') {
            if (typeof val === 'number') return val

            const isMin = type === 'min'
            let result = 0
            if (this.scatter) {
                const values = this.dataset.map(d => d[axis === 'y' ? this.axis.y.datakey : this.axis.x.datakey])
                result = isMin ? Math.min(...values) : Math.max(...values)
            } else {
                result = bound(this.curData, type, isMin ? 0 : 1)

                if (isMin && result === 0) {
                    result = bound(this.curData, 'min', 1)
                }
            }
            if (typeof val === 'function') return val(result)

            return result
        },
    },
    render (h) {
        const slots = this.$slots.default || []
        let datakeys = [] // We need get slots datakey prop to calculate max and min values for the scales
        const cartesians = [] // We need inject necessary props to cartesian slots
        const grid = []
        const others = []
        const axis = []
        const plugins = []

        // Props
        const activeElements = []
        const legends = []

        // Reset snap to manage bars
        this.snap = {}

        slots.forEach((slot) => {
            const options = slot.componentOptions
            if (!options) {
                others.push(slot)
                return
            }
            const sealed = options.Ctor.sealedOptions
            if (!sealed) {
                return
            }
            const { propsData } = options
            const { datakey, legend } = propsData
            const cartesiansLength = cartesians.length

            switch (sealed.type) {
                case 'cartesian':
                    // Add datekeys. Removed unique datakey condiition to use multiple elements
                    if (datakey) datakeys = [...datakeys.filter(key => key !== datakey), datakey]
                    // Add to actives elements
                    activeElements.push(cartesiansLength)
                    // Add to legends elements
                    if (legend) legends.push(legend)
                    // Add slot
                    slot.index = cartesiansLength
                    cartesians.push(slot)
                    break
                case 'grid':
                    grid.push(slot)
                    break
                case 'axis':
                    axis.push(slot)
                    break
                case 'plugins':
                    plugins.push(slot)
                    break
                default:
                    break
            }

            if (sealed.preload) {
                sealed.preload({ parent: this, props: propsData, index: cartesiansLength })
            }
        })

        const { viewWidth, height, responsive } = this
        this.datakeys = datakeys
        this.activeElements = activeElements
        this.legends = legends

        return h(
            'div',
            {
                class: 'WCartesian',
                style: {
                    position: 'relative',
                    width: responsive ? '100%' : `${viewWidth}px`,
                },
            },
            [
                h(
                    'svg',
                    {
                        attrs: {
                            width: viewWidth,
                            height,
                            viewBox: `0 0 ${viewWidth} ${height}`,
                        },
                    },
                    this.chartReady ? [others, grid, cartesians, axis] : []
                ),
                this.chartReady ? plugins : [],
            ]
        )
    },
}
</script>
