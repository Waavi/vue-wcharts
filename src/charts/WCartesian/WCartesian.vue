<script>
import VueTypes from 'vue-types'
import { scaleLinear } from 'd3-scale'
import stack from 'd3-shape/src/stack'
import stackOffsetDiverging from 'd3-shape/src/offset/diverging'
import noop from 'lodash.noop'
import { bound } from '../../utils/maths'
import { isNumber } from '../../utils/checks'
import merge from '../../utils/merge'
import chartMixin from '../../mixins/chart'

export default {
    name: 'WCartesian',
    mixins: [chartMixin],
    props: {
        scatter: VueTypes.bool.def(false),
        bound: VueTypes.array.def([]),
        xBound: VueTypes.array.def([]),
        gap: VueTypes.oneOfType([VueTypes.number, VueTypes.arrayOf(VueTypes.number).def([0, 0, 0, 0])]).def(0),
    },
    data () {
        return {
            axis: {
                x: {
                    datakey: null,
                    name: null,
                    numTicks: null,
                },
                y: {
                    datakey: null,
                    name: null,
                    numTicks: null,
                },
                z: {
                    datakey: null,
                    name: null,
                    range: null,
                },
            },
        }
    },
    computed: {
        // yScale with scaleLinear of d3
        // ref: https://github.com/d3/d3-scale#_continuous
        yScale () {
            const domain = [this.bounds.min, this.bounds.max]
            const bottom = this.canvas.y1 - this.padding[2]
            const top = this.canvas.y0 + this.padding[0]
            return scaleLinear()
                .domain(domain)
                .range([bottom, top])
        },
        // xScale with scaleLinear of d3
        // ref: https://github.com/d3/d3-scale#_continuous
        xScale () {
            const domain = this.scatter ? [this.xBounds.min, this.xBounds.max] : [0, this.data.length - 1]
            const left = this.canvas.x0 + this.padding[3]
            const right = this.canvas.x1 - this.padding[1]
            return scaleLinear()
                .domain(domain)
                .range([left, right])
        },
        // zScale calculate like Recharts
        zScale () {
            if (this.scatter && this.axis.z.datakey) {
                const [rangeMin, rangeMax] = this.axis.z.range
                const { min: boundMin, max: boundMax } = this.zBounds
                return val => Math.sqrt((((val - boundMin) / (boundMax - boundMin)) * (rangeMax - rangeMin) + rangeMin) / Math.PI)
            }
            return val => val
        },
        // bounds
        bounds () {
            if (this.datakeys.length || this.axis.y.datakey) {
                const [boundMin, boundMax] = this.bound
                return this.sanitizeBounds({
                    min: this.getBound(boundMin, 'min', 'y'),
                    max: this.getBound(boundMax, 'max', 'y'),
                })
            }
            return { max: 0, min: 0 }
        },
        xBounds () {
            if (this.scatter && this.axis.x.datakey) {
                const [boundMin, boundMax] = this.xBound
                return this.sanitizeBounds({
                    min: this.getBound(boundMin, 'min', 'x'),
                    max: this.getBound(boundMax, 'max', 'x'),
                })
            }
            return { max: 0, min: 0 }
        },
        zBounds () {
            if (this.scatter && this.axis.z.datakey) {
                const values = this.data.map(d => d[this.axis.z.datakey])
                return this.sanitizeBounds({
                    min: Math.min(...values),
                    max: Math.max(...values),
                })
            }
            return { max: 0, min: 0 }
        },
        // Generate padding by array or number.
        // return [10, 10, 10, 10]
        padding () {
            const gap = Array.isArray(this.gap) ? this.gap : Array(4).fill(this.gap)
            return gap.map((item, index) => item + this.offset[index])
        },
        // Offset of bars
        // return [10, 10, 10, 10]
        offset () {
            const offset = this.maxBarWidth * (this.numberOfBarsPerGroup / 2)
            const gap = Array(4).fill(0)

            if (this.numberOfBarsPerGroup) {
                gap[1] = offset
                gap[3] = offset
                return gap
            }

            return gap
        },
        stackedCurData () {
            const stackDatakeys = this.getActiveElementsDatakeys(this.snap.stackedBarsByIndex)
            return this.getDataStacked(stackDatakeys, stackOffsetDiverging)
        },
        barsCurData () {
            const stackDatakeys = Object.values(this.snap.stackedBarsByIndex || []).map(b => b.datakey)
            const barsDatakeys = this.getActiveElementsDatakeys(this.snap.barsByIndex)
                .filter(([index, value]) => !stackDatakeys.includes(value.datakey))
            return this.getDataStacked(barsDatakeys)
        },
        linesCurData () {
            const linesDatakeys = this.getActiveElementsDatakeys(this.snap.linesByIndex || [])
            return this.getDataStacked(linesDatakeys)
        },
        curData () {
            return [...this.stackedCurData, ...this.barsCurData, ...this.linesCurData]
        },
        // Return number of bars per group. If is stacked, 'numberOfBars' is one.
        numberOfBarsPerGroup () {
            const bars = this.getActiveElements(this.snap.barIds).length
            const stackedBars = this.getActiveElements(this.snap.stackedBarIds).length
            return bars - stackedBars + (stackedBars ? 1 : 0)
        },
        // Return max possible bar width saving space between group of bars
        maxBarWidth () {
            if (!this.numberOfBarsPerGroup) return 0
            return this.canvas.width / (this.numberOfBarsPerGroup * (this.data.length + 1))
        },
        // Return position of bars per group
        positionsPerGroupOfBars () {
            const stackedBars = this.getActiveElements(this.snap.stackedBarIds)
            const positions = this.getActiveElements(this.snap.barIds).filter(b => !stackedBars.includes(b))
            if (stackedBars.length) positions.splice(positions.filter(b => b < stackedBars[0]).length, 0, stackedBars)
            return positions
        },
    },
    methods: {
        // Calc min/max bound values
        getBound (val, type = 'min', axis = 'y') {
            if (typeof val === 'number') return val

            const isMin = type === 'min'
            let result = 0
            if (this.scatter) {
                const values = this.data.map(d => d[axis === 'y' ? this.axis.y.datakey : this.axis.x.datakey])
                result = isMin ? Math.min(...values) : Math.max(...values)
            } else {
                result = bound(this.curData, type, isMin ? 0 : 1)

                if (isMin && result === 0) {
                    result = bound(this.curData, 'min', 1)
                }
            }
            if (typeof val === 'function') return val(result)

            return result
        },
        sanitizeBounds ({ min, max }) {
            if (min === max) {
                if (min === 0) {
                    return { min: 0, max: 1 }
                }
                if (min < 0) {
                    return { min: -2 * min, max: 0 }
                }
                return { min: 0, max: 2 * max }
            }
            return { min, max }
        },
        // Return list of datakeys of an array of active elements (bar, stacked bars or lines)
        getActiveElementsDatakeys (elements) {
            return Object.entries(elements || [])
                .filter(([index, value]) => this.activeElements.includes(parseInt(index, 10)))
                .map(([index, value]) => value.datakey)
        },
        getActiveElements (ids) {
            return ids ? Object.values(ids).filter(b => this.activeElements.includes(b)) : []
        },
        // Return data without transform of datakeys and dataset props
        // ref: https://github.com/d3/d3-shape#stacks
        getDataStacked (datakeys, offset = noop) {
            return stack()
                .keys(datakeys)
                .value((d, key) => (isNumber(d[key]) ? d[key] : null))
                .offset(offset)(this.data)
        },
        // Set axis options
        setAxisOptions (options) {
            this.axis = merge(this.axis, options)
        },
    },
    render (h) {
        const slots = this.$slots.default || []
        let datakeys = [] // We need get slots datakey prop to calculate max and min values for the scales
        const cartesians = [] // We need inject necessary props to cartesian slots
        const grid = []
        const others = []
        const axis = []
        const plugins = []

        // Props
        const legends = []

        // Reset snap to manage bars
        this.snap = {}

        slots.forEach((slot) => {
            const options = slot.componentOptions
            if (!options) {
                others.push(slot)
                return
            }
            const sealed = options.Ctor.sealedOptions
            if (!sealed) {
                return
            }
            const { propsData } = options
            const { datakey, legend } = propsData || {}
            const cartesiansLength = cartesians.length

            switch (sealed.type) {
                case 'cartesian':
                    // Add datekeys. Removed unique datakey condiition to use multiple elements
                    if (datakey) datakeys = [...datakeys.filter(key => key !== datakey), datakey]
                    // Add to legends elements
                    if (legend) legends.push(legend)
                    // Add slot
                    slot.componentOptions.propsData.index = cartesiansLength
                    cartesians.push(slot)
                    break
                case 'grid':
                    grid.push(slot)
                    break
                case 'axis':
                    axis.push(slot)
                    break
                case 'plugins':
                    plugins.push(slot)
                    break
                default:
                    break
            }

            if (sealed.preload) {
                sealed.preload({ parent: this, props: propsData, index: cartesiansLength })
            }
        })

        const { viewWidth, height, responsive } = this
        this.datakeys = datakeys
        this.legends = legends

        return h(
            'div',
            {
                class: 'WCartesian',
                style: {
                    position: 'relative',
                    width: responsive ? '100%' : `${viewWidth}px`,
                },
            },
            [
                h(
                    'svg',
                    {
                        attrs: {
                            width: viewWidth,
                            height,
                            viewBox: `0 0 ${viewWidth} ${height}`,
                        },
                    },
                    this.chartReady ? [others, grid, cartesians, axis] : []
                ),
                this.chartReady ? plugins : [],
            ]
        )
    },
}
</script>
