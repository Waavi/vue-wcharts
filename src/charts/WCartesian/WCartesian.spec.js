import { shallowMount, mount } from '@vue/test-utils'
import localVue from '../../../config/tests'
import WCartesian from './WCartesian.vue'
import {
    WXAxis,
    WYAxis,
    // WBar,
    WCartesianGrid,
    // WLine,
    WLegend,
    // WMarker,
    WScatter,
} from '../../components'

describe('Charts/WCartesian', () => {
    const dataset = [
        {
            name: 'Page A',
            one: 1000,
            two: 2400,
            three: 2400,
        },
        {
            name: 'Page B',
            one: 3000,
            two: 1398,
            three: 2210,
        },
        {
            name: 'Page C',
            one: 2000,
            two: 9800,
            three: 2290,
        },
    ]

    const axisSlots = {
        default: [WXAxis, WYAxis],
    }

    const legendAndGridSlots = {
        default: [WCartesianGrid, WLegend],
    }

    const scatterSlots = {
        default: [WScatter],
    }
    const propsData = {
        dataset,
    }

    const defaultConfig = {
        localVue,
        propsData,
    }

    it(`Should be render correctly`, () => {
        const wrapper = shallowMount(WCartesian, defaultConfig)
        expect(wrapper.html()).toMatchSnapshot()
    })

    it(`Should be render correctly with axix`, () => {
        const wrapper = shallowMount(WCartesian, { ...defaultConfig, slots: axisSlots })
        expect(wrapper.html()).toMatchSnapshot()
    })

    it(`Should be render correctly, with gap it is a number and array of numbers`, () => {
        const wrapperNumber = shallowMount(WCartesian, {
            ...defaultConfig,
            propsData: {
                ...defaultConfig.propsData,
                gap: 5,
            },
            slots: axisSlots,
        })
        expect(wrapperNumber.html()).toMatchSnapshot()

        const wrapperArray = shallowMount(WCartesian, {
            ...defaultConfig,
            propsData: {
                ...defaultConfig.propsData,
                gap: [5, 5, 5, 5],
            },
            slots: axisSlots,
        })
        expect(wrapperArray.html()).toMatchSnapshot()
    })

    it(`Should be render correctly with bars and lines`, () => {
        const wrapper = shallowMount(WCartesian, {
            ...defaultConfig,
            slots: {
                default: `
                    <WBar id="testBar" datakey="one" />
                    <WLine datakey="one" legend="One Line" />
                    <WLegend position="top" align="end" selectable />
                `,
            },
        })
        expect(wrapper.html()).toMatchSnapshot()
    })

    it(`Should be render correctly with legend and markers`, () => {
        const wrapper = shallowMount(WCartesian, { ...defaultConfig, slots: legendAndGridSlots })
        expect(wrapper.html()).toMatchSnapshot()
    })

    it(`Should be render correctly with scatter`, () => {
        const wrapper = shallowMount(WCartesian, {
            ...defaultConfig,
            propsData: { ...defaultConfig.propsData, scatter: true },
            slots: scatterSlots,
        })
        expect(wrapper.html()).toMatchSnapshot()
    })

    it(`Should be return bound ox x axis`, () => {
        const wrapper = mount(WCartesian, { ...defaultConfig })
        const bound = wrapper.vm.getBound('10', 'min')
        expect(bound).toEqual(0)

        const boundMax = wrapper.vm.getBound(n => n + 1000, 'max')
        expect(boundMax).toEqual(1000)
    })

    it(`Should be calc bound if execute sanitizeBounds method correctly`, () => {
        const wrapper = shallowMount(WCartesian, { ...defaultConfig, slots: axisSlots })
        expect(wrapper.vm.sanitizeBounds({ min: 0, max: 0 })).toEqual({ min: 0, max: 1 })
        expect(wrapper.vm.sanitizeBounds({ min: -1, max: -1 })).toEqual({ min: 2, max: 0 })
        expect(wrapper.vm.sanitizeBounds({ min: 2, max: 2 })).toEqual({ min: 0, max: 4 })
    })
})
