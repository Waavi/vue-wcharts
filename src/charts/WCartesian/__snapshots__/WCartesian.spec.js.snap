// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Charts/WCartesian Should be render correctly 1`] = `<div class="WCartesian" style="position: relative; width: 600px;"><svg width="600" height="400" viewBox="0 0 600 400"></svg></div>`;

exports[`Charts/WCartesian Should be render correctly with axix 1`] = `
<div class="WCartesian" style="position: relative; width: 600px;"><svg width="600" height="400" viewBox="0 0 600 400">
    <g id="WXAxis">
      <line x1="40" y1="376" x2="580" y2="376" stroke="#cccccc"></line>
      <g text-anchor="middle">
        <line x1="40" y1="376" x2="40" y2="381" stroke="#cccccc"></line>
        <WTickText-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#6e19ff,#dfbd46" x="40" y="396" dy="" index="0" value="0" styles="[object Object]"></WTickText-stub>
      </g>
      <g text-anchor="middle">
        <line x1="310" y1="376" x2="310" y2="381" stroke="#cccccc"></line>
        <WTickText-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#6e19ff,#dfbd46" x="310" y="396" dy="" index="1" value="1" styles="[object Object]"></WTickText-stub>
      </g>
      <g text-anchor="middle">
        <line x1="580" y1="376" x2="580" y2="381" stroke="#cccccc"></line>
        <WTickText-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#6e19ff,#dfbd46" x="580" y="396" dy="" index="2" value="2" styles="[object Object]"></WTickText-stub>
      </g>
      <!---->
      <!---->
    </g>
    <g id="WYAxis">
      <line x1="40" y1="10" x2="40" y2="376" stroke="#cccccc"></line>
      <g text-anchor="end">
        <line x1="35" y1="193" x2="40" y2="193" stroke="#cccccc"></line>
        <WTickText-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#6e19ff,#dfbd46" x="30" y="197" dy="" index="0" value="0" styles="[object Object]"></WTickText-stub>
      </g>
      <g text-anchor="end">
        <line x1="35" y1="193" x2="40" y2="193" stroke="#cccccc"></line>
        <WTickText-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#6e19ff,#dfbd46" x="30" y="197" dy="" index="1" value="0" styles="[object Object]"></WTickText-stub>
      </g>
      <g text-anchor="end">
        <line x1="35" y1="193" x2="40" y2="193" stroke="#cccccc"></line>
        <WTickText-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#6e19ff,#dfbd46" x="30" y="197" dy="" index="2" value="0" styles="[object Object]"></WTickText-stub>
      </g>
      <g text-anchor="end">
        <line x1="35" y1="193" x2="40" y2="193" stroke="#cccccc"></line>
        <WTickText-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#6e19ff,#dfbd46" x="30" y="197" dy="" index="3" value="0" styles="[object Object]"></WTickText-stub>
      </g>
      <g text-anchor="end">
        <line x1="35" y1="193" x2="40" y2="193" stroke="#cccccc"></line>
        <WTickText-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#6e19ff,#dfbd46" x="30" y="197" dy="" index="4" value="0" styles="[object Object]"></WTickText-stub>
      </g>
      <g text-anchor="end">
        <line x1="35" y1="193" x2="40" y2="193" stroke="#cccccc"></line>
        <WTickText-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#6e19ff,#dfbd46" x="30" y="197" dy="" index="5" value="0" styles="[object Object]"></WTickText-stub>
      </g>
      <g text-anchor="end">
        <line x1="35" y1="193" x2="40" y2="193" stroke="#cccccc"></line>
        <WTickText-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#6e19ff,#dfbd46" x="30" y="197" dy="" index="6" value="0" styles="[object Object]"></WTickText-stub>
      </g>
      <g text-anchor="end">
        <line x1="35" y1="193" x2="40" y2="193" stroke="#cccccc"></line>
        <WTickText-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#6e19ff,#dfbd46" x="30" y="197" dy="" index="7" value="0" styles="[object Object]"></WTickText-stub>
      </g>
      <!---->
    </g>
  </svg></div>
`;

exports[`Charts/WCartesian Should be render correctly with legend and markers 1`] = `
<div class="WCartesian" style="position: relative; width: 600px;"><svg width="600" height="400" viewBox="0 0 600 400">
    <g>
      <line x1="0" y1="200" x2="600" y2="200" stroke="#cccccc" stroke-width="1" stroke-dasharray="3" fill="none"></line>
      <line x1="0" y1="200" x2="600" y2="200" stroke="#cccccc" stroke-width="1" stroke-dasharray="3" fill="none"></line>
      <line x1="0" y1="200" x2="600" y2="200" stroke="#cccccc" stroke-width="1" stroke-dasharray="3" fill="none"></line>
      <line x1="0" y1="200" x2="600" y2="200" stroke="#cccccc" stroke-width="1" stroke-dasharray="3" fill="none"></line>
      <line x1="0" y1="200" x2="600" y2="200" stroke="#cccccc" stroke-width="1" stroke-dasharray="3" fill="none"></line>
      <line x1="0" y1="200" x2="600" y2="200" stroke="#cccccc" stroke-width="1" stroke-dasharray="3" fill="none"></line>
      <line x1="0" y1="200" x2="600" y2="200" stroke="#cccccc" stroke-width="1" stroke-dasharray="3" fill="none"></line>
      <line x1="0" y1="200" x2="600" y2="200" stroke="#cccccc" stroke-width="1" stroke-dasharray="3" fill="none"></line>
      <line x1="0" y1="0" x2="0" y2="400" stroke="#cccccc" stroke-width="1" stroke-dasharray="3" fill="none"></line>
      <line x1="85.71428571428571" y1="0" x2="85.71428571428571" y2="400" stroke="#cccccc" stroke-width="1" stroke-dasharray="3" fill="none"></line>
      <line x1="171.42857142857142" y1="0" x2="171.42857142857142" y2="400" stroke="#cccccc" stroke-width="1" stroke-dasharray="3" fill="none"></line>
      <line x1="257.1428571428571" y1="0" x2="257.1428571428571" y2="400" stroke="#cccccc" stroke-width="1" stroke-dasharray="3" fill="none"></line>
      <line x1="342.85714285714283" y1="0" x2="342.85714285714283" y2="400" stroke="#cccccc" stroke-width="1" stroke-dasharray="3" fill="none"></line>
      <line x1="428.57142857142856" y1="0" x2="428.57142857142856" y2="400" stroke="#cccccc" stroke-width="1" stroke-dasharray="3" fill="none"></line>
      <line x1="514.2857142857142" y1="0" x2="514.2857142857142" y2="400" stroke="#cccccc" stroke-width="1" stroke-dasharray="3" fill="none"></line>
      <line x1="600" y1="0" x2="600" y2="400" stroke="#cccccc" stroke-width="1" stroke-dasharray="3" fill="none"></line>
    </g>
  </svg>
  <!---->
</div>
`;

exports[`Charts/WCartesian Should be render correctly with scatter 1`] = `
<div class="WCartesian" style="position: relative; width: 600px;"><svg width="600" height="400" viewBox="0 0 600 400">
    <g>
      <!---->
      <g>
        <WDot-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#6e19ff,#dfbd46" index="0" cartesianIndex="0" trigger="hover" x="0" y="0" fill="#48c0b6" stroke="#48c0b6" strokeWidth="0" radius="8" hoverRadius="8" info="[object Object]" styles="[object Object]" transition="all 250ms ease" z="8" opacity="0.8"></WDot-stub>
        <WDot-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#6e19ff,#dfbd46" index="1" cartesianIndex="0" trigger="hover" x="0" y="0" fill="#48c0b6" stroke="#48c0b6" strokeWidth="0" radius="8" hoverRadius="8" info="[object Object]" styles="[object Object]" transition="all 250ms ease" z="8" opacity="0.8"></WDot-stub>
        <WDot-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#6e19ff,#dfbd46" index="2" cartesianIndex="0" trigger="hover" x="0" y="0" fill="#48c0b6" stroke="#48c0b6" strokeWidth="0" radius="8" hoverRadius="8" info="[object Object]" styles="[object Object]" transition="all 250ms ease" z="8" opacity="0.8"></WDot-stub>
      </g>
    </g>
  </svg></div>
`;
